src/Parser/BrewmasterMonk/Modules/Features/Energy.js

```diff
@@ -0,0 +1,125 @@
+    this.baseHaste = this.combatants.selected.hastePercentage;
+    this.lastChangeInEnergy = this.owner.fight.start_time;
+    this.fightEnd = this.owner.fight.end_time;
+    this.currentRegen = this.calcEnergyRegen([this.baseHaste], this.lastChangeInEnergy);
+  }
+
+  // Multiplicative haste bonuses by putting each one in an array and reducing it
+  calcEnergyRegen(hasteArray, timestamp) {
+    const addedEnergy = ((timestamp - this.lastChangeInEnergy) / 1000) * this.currentRegen;
+    this.lastChangeInEnergy = timestamp;
+    this.totalEnergyRegen += addedEnergy || 0;
+    if (typeof hasteArray === 'object' && hasteArray.reduce) {
+      const totalHaste = hasteArray.reduce((total, haste) => total *= (haste + 1), 1);
+      return BASE_REGEN * totalHaste;
+    }      
+    return BASE_REGEN;
+  }  
+  
+  getEnergyAtTime(timestamp) {
+    const lastEvent = this.energy[this.energy.length - 1];
+    const timeSinceLastAbilityinSecs = (timestamp - lastEvent.timestamp) / 1000;
+    const addedEnergy = Math.round(timeSinceLastAbilityinSecs * this.currentRegen);
+    const energyAtLastAbility = lastEvent.amount - lastEvent.cost;
+    return energyAtLastAbility + addedEnergy;
+  }
+
+  calcHowLongAtMax(timestamp) {
+    // beginning of the fight you likely start at 100
+    if (this.energy.length > 0) {
+      const lastEvent = this.energy[this.energy.length - 1];
+      const timeSinceLastAbilityinSecs = (timestamp - lastEvent.timestamp) / 1000;
+      const energyAtLastAbility = lastEvent.amount - lastEvent.cost;
+      const timeToMax = (MAX_ENERGY - energyAtLastAbility) / this.currentRegen;
+      return (timeSinceLastAbilityinSecs - timeToMax);
+    }
+    return 0;
+  }
+
+  addEnergyEvent(event, energy, cost) {
+    this.energy.push({
+      timestamp: event.timestamp, 
+      amount: energy,
+      cost: cost,
+    });
+  }
+
+  on_byPlayer_cast(event) {
+    const spellId = event.ability.guid;
+    if (SPELLS.BLACK_OX_BREW_TALENT.id === spellId) {
+      const currentEnergy = this.getEnergyAtTime(event.timestamp);
+      this.energyWastedFromBOB += (100 - currentEnergy);
+      this.addEnergyEvent(event, 100, 0);
+      this.bobCasts++;
+    }
+    if (SPELLS_WITH_ENERGY_COST[spellId] !== undefined) {
+      const currentEnergy = event.classResources.filter(a => a.type === RESOURCE_TYPES.ENERGY)[0].amount;
+      if (currentEnergy === MAX_ENERGY) {
+        const time = this.calcHowLongAtMax(event.timestamp);
+        if (time > 0) {
+          this.timeAtMaxEnergy += time;
+          this.energyWastedFromMax += time * this.currentRegen;
+        }
+      }
+      this.addEnergyEvent(event, currentEnergy, SPELLS_WITH_ENERGY_COST[spellId]);
+    }
+  }
+
+  on_byPlayer_applybuff(event) {
+    const spellId = event.ability.guid;
+    if (SPELLS_WITH_HASTE[spellId] !== undefined) {
+      debug && console.log('hero');
+      this.currentRegen = this.calcEnergyRegen([this.baseHaste, SPELLS_WITH_HASTE[spellId] / 100], event.timestamp);
```

Comment:
This only works with *one* ability at a time and this code is kinda hard to follow. Could this be refactored a bit? I think if you made `calcEnergyRegen` either a completely pure method or without any return values it would already be a fair bit easier to follow.

If you're interested in the currentHaste it might be best to put a dependency on AlwaysBeCasting and use its `currentHaste` property.